File: benchmarks/pandas/bench_sql.py
"""Benchmark: read_sql / to_sql on 10k-row DataFrames using SQLite in-memory"""
import json
import time
import math
import sqlite3
import pandas as pd

ROWS = 10_000
WARMUP = 3
ITERATIONS = 20

# ── Build a matching dataset ──────────────────────────────────────────────────
data = {
    "id": list(range(ROWS)),
    "value": [math.sin(i * 0.01) * 1000 for i in range(ROWS)],
    "label": [f"item_{i % 100}" for i in range(ROWS)],
}
df = pd.DataFrame(data)

# ── SQLite in-memory database ─────────────────────────────────────────────────
con = sqlite3.connect(":memory:")
df.to_sql("mock_table", con, index=False, if_exists="replace")

# ── Warm-up reads ─────────────────────────────────────────────────────────────
for _ in range(WARMUP):
    pd.read_sql_query("SELECT * FROM mock_table", con)

# ── read_sql_query benchmark ──────────────────────────────────────────────────
start_read = time.perf_counter()
for _ in range(ITERATIONS):
    pd.read_sql_query("SELECT * FROM mock_table", con)
total_read = (time.perf_counter() - start_read) * 1000

# ── Warm-up writes ────────────────────────────────────────────────────────────
for _ in range(WARMUP):
    df.to_sql("bench_table", con, index=False, if_exists="replace")

# ── to_sql benchmark ──────────────────────────────────────────────────────────
start_write = time.perf_counter()
for _ in range(ITERATIONS):
    df.to_sql("bench_table", con, index=False, if_exists="replace")
total_write = (time.perf_counter() - start_write) * 1000

con.close()

print(json.dumps({
    "function": "sql",
    "mean_ms": (total_read + total_write) / (2 * ITERATIONS),
    "iterations": ITERATIONS,
    "total_ms": total_read + total_write,
    "read_mean_ms": total_read / ITERATIONS,
    "write_mean_ms": total_write / ITERATIONS,
}))

File: benchmarks/tsb/bench_sql.ts
/**
 * Benchmark: readSqlQuery / toSql on 10k-row result sets
 */
import { DataFrame, readSqlQuery, toSql } from "../../src/index.js";
import type { SqlConnection, SqlResult, SqlRow, SqlValue, IfExistsStrategy } from "../../src/index.js";

const ROWS = 10_000;
const WARMUP = 3;
const ITERATIONS = 20;

// ── Shared result set ─────────────────────────────────────────────────────────
const columns: string[] = ["id", "value", "label"];
const rows: SqlRow[] = Array.from({ length: ROWS }, (_, i) => ({
  id: i,
  value: Math.sin(i * 0.01) * 1000,
  label: `item_${i % 100}`,
}));

// ── Mock adapter for reads ────────────────────────────────────────────────────
class ReadAdapter implements SqlConnection {
  query(_sql: string, _params?: readonly SqlValue[]): SqlResult {
    return { columns, rows };
  }
  listTables(): readonly string[] {
    return ["mock_table"];
  }
}

const readConn = new ReadAdapter();

// ── Warm-up reads ─────────────────────────────────────────────────────────────
for (let i = 0; i < WARMUP; i++) {
  readSqlQuery("SELECT * FROM mock_table", readConn);
}

// ── readSqlQuery benchmark ────────────────────────────────────────────────────
const startRead = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  readSqlQuery("SELECT * FROM mock_table", readConn);
}
const totalRead = performance.now() - startRead;

// ── Mock adapter for writes ───────────────────────────────────────────────────
class WriteAdapter implements SqlConnection {
  query(_sql: string, _params?: readonly SqlValue[]): SqlResult {
    return { columns: [], rows: [] };
  }
  listTables(): readonly string[] {
    return [];
  }
  insert(
    _tableName: string,
    _rows: readonly SqlRow[],
    _columns: readonly string[],
    _ifExists: IfExistsStrategy,
  ): number {
    return _rows.length;
  }
}

const writeConn = new WriteAdapter();
const df = readSqlQuery("SELECT * FROM mock_table", readConn);

// ── Warm-up writes ────────────────────────────────────────────────────────────
for (let i = 0; i < WARMUP; i++) {
  toSql(df, "bench_table", writeConn, { ifExists: "replace" });
}

// ── toSql benchmark ───────────────────────────────────────────────────────────
const startWrite = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  toSql(df, "bench_table", writeConn, { ifExists: "replace" });
}
const totalWrite = performance.now() - startWrite;

console.log(
  JSON.stringify({
    function: "sql",
    mean_ms: (totalRead + totalWrite) / (2 * ITERATIONS),
    iterations: ITERATIONS,
    total_ms: totalRead + totalWrite,
    read_mean_ms: totalRead / ITERATIONS,
    write_mean_ms: totalWrite / ITERATIONS,
  }),
);
